integration/sandbox_run_rollback_test.go

```diff
@@ -89,6 +89,150 @@ func TestRunPodSandboxWithShimStartFailure(t *testing.T) {
 	require.Equal(t, true, strings.Contains(err.Error(), "no hard feelings"))
 }
 
+// TestRunPodSandboxWithShimDeleteFailure should keep the sandbox record if
+// failed to rollback shim by shim.Delete API.
+func TestRunPodSandboxWithShimDeleteFailure(t *testing.T) {
+	if runtime.GOOS != "linux" {
+		t.Skip()
+	}
+	if os.Getenv("ENABLE_CRI_SANDBOXES") == "" {
+		t.Skip()
+	}
+
+	testCase := func(restart bool) func(*testing.T) {
+		return func(t *testing.T) {
+			t.Logf("Init PodSandboxConfig with specific label")
+			labels := map[string]string{
+				t.Name(): "true",
+			}
+			sbConfig := PodSandboxConfig(t.Name(), "failpoint", WithPodLabels(labels))
+
+			t.Logf("Inject Shim failpoint")
+			injectShimFailpoint(t, sbConfig, map[string]string{
+				"Start":  "1*error(failed to start shim)",
+				"Delete": "1*error(please retry)", // inject failpoint during rollback shim
+			})
+
+			t.Logf("Create a sandbox")
+			_, err := runtimeService.RunPodSandbox(sbConfig, failpointRuntimeHandler)
+			require.Error(t, err)
+			require.Equal(t, true, strings.Contains(err.Error(), "failed to start shim"))
+
+			t.Logf("ListPodSandbox with the specific label")
+			l, err := runtimeService.ListPodSandbox(&criapiv1.PodSandboxFilter{LabelSelector: labels})
+			require.NoError(t, err)
+			require.Equal(t, true, len(l) == 1)
+
+			sb := l[0]
+			require.Equal(t, true, sb.State == criapiv1.PodSandboxState_SANDBOX_NOTREADY)
+			require.Equal(t, true,
+				sb.Metadata.Name == sbConfig.Metadata.Name &&
+					sb.Metadata.Namespace == sbConfig.Metadata.Namespace &&
+					sb.Metadata.Uid == sbConfig.Metadata.Uid &&
+					sb.Metadata.Attempt == sbConfig.Metadata.Attempt,
+			)
+
+			t.Logf("Check PodSandboxStatus")
+			sbStatus, err := runtimeService.PodSandboxStatus(sb.Id)
+			require.NoError(t, err)
+			require.Equal(t, true, sbStatus.State == criapiv1.PodSandboxState_SANDBOX_NOTREADY)
+			require.Equal(t, true, len(sbStatus.Network.Ip) != 0)
+
+			if restart {
+				t.Logf("Restart containerd")
+				RestartContainerd(t)
+
+				t.Logf("ListPodSandbox with the specific label")
+				l, err = runtimeService.ListPodSandbox(&criapiv1.PodSandboxFilter{Id: sb.Id})
+				require.NoError(t, err)
+				require.Equal(t, true, len(l) == 1)
+				require.Equal(t, true, l[0].State == criapiv1.PodSandboxState_SANDBOX_NOTREADY)
+
+				t.Logf("Check PodSandboxStatus")
+				sbStatus, err := runtimeService.PodSandboxStatus(sb.Id)
+				fmt.Println(sbStatus.Network)
```

Comment:
remove this line or replace it with t.Logf

Reply:
fixed.